zookeeper-api/src/test/java/org/apache/helix/zookeeper/impl/client/TestZkClientAsyncFailure.java

```diff
@@ -0,0 +1,207 @@
+
+  @BeforeClass
+  public void beforeClass() throws JMException {
+    _zkClient = _zkServerMap.values().iterator().next().getZkClient();
+    _zkServerAddress = _zkClient.getServers();
+    _zkClient.createPersistent(TEST_ROOT);
+
+    _monitor = new ZkClientMonitor(TEST_TAG, TEST_KEY, TEST_INSTANCE, false, null);
+    _monitor.register();
+  }
+
+  @AfterClass
+  public void afterClass() {
+    _monitor.unregister();
+    _zkClient.deleteRecursively(TEST_ROOT);
+    _zkClient.close();
+  }
+
+  private ObjectName buildObjectName(String tag, String key, String instance)
+      throws MalformedObjectNameException {
+    return ZkClientMonitor.getObjectName(tag, key, instance);
+  }
+
+  private ObjectName buildPathMonitorObjectName(String tag, String key, String instance,
+      String path) throws MalformedObjectNameException {
+    return new ObjectName(String.format("%s,%s=%s", buildObjectName(tag, key, instance).toString(),
+        ZkClientPathMonitor.MONITOR_PATH, path));
+  }
+
+  @Test
+  public void testAsyncWrite() throws JMException {
+    TestZkClientAsyncFailure.MockAsyncZkClient testZkClient =
+        new TestZkClientAsyncFailure.MockAsyncZkClient(_zkServerAddress);
+    try {
+      ZNRecord tmpRecord = new ZNRecord("tmpRecord");
+      tmpRecord.setSimpleField("foo", "bar");
+      testZkClient.createPersistent(NODE_PATH, tmpRecord);
+
+      ObjectName instancesName = buildPathMonitorObjectName(TEST_TAG, TEST_KEY, TEST_INSTANCE,
+          ZkClientPathMonitor.PredefinedPath.Root.name());
+
+      ZkAsyncCallbacks.SetDataCallbackHandler setCallback =
+          new ZkAsyncCallbacks.SetDataCallbackHandler();
+      Assert.assertEquals(setCallback.getRc(), UNKNOWN_RET_CODE);
+
+      tmpRecord.setSimpleField("test", "data");
+
+      // asyncSet should succeed because the return code is OK
+      testZkClient.setAsyncCallRC(KeeperException.Code.OK.intValue());
+      testZkClient.asyncSetData(NODE_PATH, tmpRecord, -1, setCallback);
+      setCallback.waitForSuccess();
+      Assert.assertEquals(setCallback.getRc(), KeeperException.Code.OK.intValue());
+      Assert.assertEquals(((ZNRecord) testZkClient.readData(NODE_PATH)).getSimpleField("test"),
+          "data");
+      Assert.assertEquals((long) _beanServer.getAttribute(instancesName,
+          ZkClientPathMonitor.PredefinedMetricDomains.WriteFailureCounter.toString()), 0);
+
+      // asyncSet should fail because the return code is APIERROR
+      testZkClient.setAsyncCallRC(KeeperException.Code.APIERROR.intValue());
+      testZkClient.asyncSetData(NODE_PATH, tmpRecord, -1, setCallback);
+      setCallback.waitForSuccess();
+      Assert.assertEquals(setCallback.getRc(), KeeperException.Code.APIERROR.intValue());
+      Assert.assertEquals((long) _beanServer.getAttribute(instancesName,
+          ZkClientPathMonitor.PredefinedMetricDomains.WriteFailureCounter.toString()), 1);
+    } finally {
+      testZkClient.close();
+      _zkClient.delete(NODE_PATH);
```

Comment:
This `_zkClient` might not be the one you want to use, because the test suite could start 3 zk servers. If testZkClient connects address1, and _zkClient connects to address2 (because you get it from the map), this cleanup won't work.

I suggest you may want to create a zkclient connecting to the `ZK_ADDR` and use this client for the whole test class. Then close it afterClass

Reply:
This makes sense, thanks. But is this a different scenario than for instance `TestZkClientAsyncRetry.testAsyncRetryCategories()`?